### ISSUE TITLE
Allow creating full stand-alone C# programs in a single .cs

### ISSUE THREAD

**Issue by mythz** [dotnet/csharplang#3502]
To build on the very exciting [C# 9 top-level programs feature](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#top-level-programs) it would be objectively awesome if in the same entry point .cs source file with the top-level program could also inline the MSBuild project information inside the same source file, so that full stand-alone programs could be written in a single .cs, e.g:

```csharp
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="AwesomeSauce" Version="1.*" />
  </ItemGroup>
</Project>

System.Console.WriteLine("Hello World!");
```

Which if there's only a single .cs with this, could be run implicitly with:

    $ dotnet run

Or explicitly be able to run multiple stand-alone scripts with:

    $ dotnet run main.cs

The beauty of this approach is it doesn't require any additional learning, you should just be able to paste the MSBuild `<Project/>` into a .cs file and have it run as if it were in 2 separate files as it's done today, I'd also expect being able to enable the existing syntax highlighting and authoring experience in .csproj files as done in [Vue's SFCs](https://vuejs.org/v2/guide/single-file-components.html). 

Ideally this would also work in **dotnet watch** which makes it vastly more enjoyable to develop  scripts when you're able to get instant feedback on every save point. 

> Please see [#Script Scripts](https://sharpscript.net/docs/sharp-scripts#live-shell-scripting-preview) for nice UX examples to create .NET Core scripts w/ real-time feedback

I'm aware of [dotnet script](https://github.com/filipw/dotnet-script), but this approach would enable full uncompromised C# programs that can do anything without anyone needing to learn the difference & limitations of dotnet script's. This would be a real C# program that you can start small and organically refactor out into into separate files as it grows in functionality.

## Bonus point rounds

Some other ideas to make the UX for developing Console Apps in C#:

### Unix friendly scripts

Support unix friendly [bash scripts](https://en.wikipedia.org/wiki/Shebang_(Unix))

```csharp
#!/usr/bin/dotnet
<Project />
// C#
```

So we're able to run C# Console Apps directly:

    $ ./main.cs

### Top-level functions

The verbosity `System.Console.WriteLine()` is starting to show its age with unnecessary distracting boilerplate with its wrist-unfriendly syntax. Top-level programs would be much nicer if we could define "top-level functions" that could be defined in System or imported from a NuGet package which could define "global" top-level functions:

```csharp
public static void print(string msg) => System.Console.WriteLine(msg);
```

So that C# programs could instead use the much more natural and approachable:

```csharp
print("Hello World!");
```

**Comment by HaloFour** [comment]
I really don't like the idea of trying to take the MSBuild syntax and trying to incorporate it into the C# language and specification.  These are completely orthogonal tools.  If you were going to key external tooling off of some kind of embedded metadata like this I would suggest using a block comment or similar.  However, IMO, trying to shove everything into a single file like this doesn't make for a "simple program", and you'd be better off using the kind of directives that CSX adds which enable a more script-like and interactive experience.

**Comment by mythz** [comment]
> If you were going to key external tooling off of some kind of embedded metadata like this I would suggest using a block comment or similar. 

@HaloFour Yep I'd be ok with a  [CDATA block](https://en.wikipedia.org/wiki/CDATA) like block for embedding MSBuild XML, e.g:

```xml
<![CDATA[
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="AwesomeSauce" Version="1.*" />
  </ItemGroup>
</Project>
]>

System.Console.WriteLine("Hello World!");
```

Still accomplishes same goal of a single encapsulated `.cs` and allows easy refacor to upgrade into a first-class dotnet project.

> However, IMO, trying to shove everything into a single file like this doesn't make for a "simple program"

It may not be "simple" for implementors, but if C# devs can reuse all their existing knowledge they know to read/author MSBuild & C# and their only perceived difference is that it's in a combined `.cs` file, then this approach would be simpler for them vs having to learn about different scripting modes & tools to execute C#.

Anyway it's very possible this approach is too radical and that using `#r` to declare inline package references many be the less drastic/preferred approach, could still accomplish a lot in a .cs with just package reference support, then there could be a command to convert a C# .cs with `#r` into an MS Build project.

**Comment by miloush** [comment]
XML comments would feel like the natural fit with no language changes required, but I am not in favour of supporting it. If you need a build file, then add a build file. Otherwise you should be happy with the defaults the compiler chooses for you (or supply it via command line).

The only reasonable suggestion is to allow the REPL syntax for referencing assemblies, though I think if that is allowed it should be allowed in any C# project.

However, if all you need is a single file, that is already possible, e.g.:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <InlineCode>

     class Program
     {
         static void Main() 
         {
             System.Console.WriteLine("Hello world!"); 
         }
     }

    </InlineCode>
  </PropertyGroup>

  <UsingTask TaskName="InlineCompile" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll" >
   <ParameterGroup>
    <Code Required="true" />
   </ParameterGroup>
   <Task>
    <Code Type="Fragment" Language="cs">
     File.WriteAllText("InlineCompile.cs", Code);
    </Code>
   </Task>
  </UsingTask> 

  <Target Name="InlineCompileTarget" BeforeTargets="BeforeBuild">
   <InlineCompile Code="$(InlineCode)" />
  </Target> 

</Project>
```

**Comment by mythz** [comment]
Embedded C# script in MSBuild with custom tasks requiring deeper knowledge with how MS Build works is far more daunting for new comers, not less, and goes against the purpose and design goals of top-level programs, to make C# more approachable. 

This basically fails at every metric, worse for new devs, worse for scripting, increases required knowledge which is non-transferable, more verbose, requires additional artificial machinery, no tooling, static analysis, inhibits refactoring into standard project. 

It's a basic failure all-round, the only purpose for embedded C# in MSBuild is to enable functionality in MSBuild that's not possible without it and why it's not used for anything else.

**Comment by CyrusNajmabadi** [comment]
> Embedded C# script in MSBuild with custom tasks requiring deeper knowledge with how MS Build works is far more daunting for new comers, not less, and goes against the purpose and design goals of top-level programs, to make C# more approachable.

Embedded MSBuild in C#, requiring deeper knowledge with how msbuild works, is far more daunting for new comers, not less, and goes against the purpose and design goals of top-level programs, to make C# more approachable.

> This basically fails at every metric, worse for new devs, worse for scripting, increases required knowledge which is non-transferable, more verbose, requires additional artificial machinery, no tooling, static analysis, inhibits refactoring into standard project.

THe same seems to be true for embedded msbuild into c#.  You need increased knowledge.  It's more verbose.  It required additional machinery.  There's no tooling or static analysis.  etc. etc.

**Comment by mythz** [comment]
> THe same seems to be true for embedded msbuild into c#. You need increased knowledge. It's more verbose. It required additional machinery. There's no tooling or static analysis. etc. etc.

How did you ever come to that conclusion? Every point you've reiterated is false. There's minimal knowlege required, you're literally pasting the MS Build project file into `.cs`. How can it be more verbose when it's 2 files concatenated together, thus saving a file node on the file system. It's much easier to provide static anlaysis when it's the exact same format / syntax used. Go look at all SPA formats that combine multiple DSL's into a singe file and still maintain their tooling and IDE support.

**Comment by mythz** [comment]
Well I can tell this posting here is a complete waste of time, every suggestion & feedback here has been worse from the actual goals of the feature request which was to make top-level programs more capable. Instead every suggestion makes the UX worse, I'm assuming commentators were against that language feature as well, since they can't see the purpose and benefits it provides. 

Anyway If the dotnet team want to implement `#r` like dotnet script & dotnet interactive that supports a watched mode I'll consider using that otherwise I'll continue using the far greater UX available in creating watched #Script Scripts.

**Comment by CyrusNajmabadi** [comment]
> There's minimal knowlege required, you're literally pasting the MS Build project file into .cs

I need the knowledge to know how that works.  To understand the interplay between teh two languages.  etc. etc.

> How can it be more verbose when it's 2 files concatenated together, thus saving a file node on the file system

Then the same was true of embedding c# into the msbuild task.  But you claimed it was more verbose even though it also saved the file on the file system :)

furthermore, saving files is a non-goal.  why would i possibly ever care about that.  2 files on my disk is *nothing*.  

>  It's much easier to provide static anlaysis when it's the exact same format / syntax used

It's literally not hte exact same format/syntax.  It's two entirely different syntaxes merged together. 

> Well I can tell posting here is a complete waste of time

Just because people don't like your idea and bring up issues with it does not mean that their issues are wrong.  C#  has a massive ecosystem with millions of developers.  As you've seen, it's challenging creating feature that people want.

TLDR: welcome to language design :)

**Comment by mythz** [comment]
@CyrusNajmabadi You're being especially obtuse and seemingly have no clue of what's being requested.

> I need the knowledge to know how that works. To understand the interplay between teh two languages. etc. etc.

No additional knowledge is required over what's currently required, if you don't need MSBuild features then you don't need the MS Build Project. Currently every dotnet project requires MSBuild.

> Then the same was true of embedding c# into the msbuild task. But you claimed it was more verbose even though it also saved the file on the file system :)

Are you being disingenuous on purpose? Count the lines my code sample is 11 LOC whilst the C# in MSBuild suggestion is 3x that.

> furthermore, saving files is a non-goal. why would i possibly ever care about that. 2 files on my disk is nothing.

Firstly I've opened a feature request with my desired goals, your desired goals are irrelevant here, if you want to propose a feature open up your own feature request, I promise not to waste your time and pollute it with worse irrelevant suggestions & false equivalences. 

Secondly, it's obviously not about saving files & disk space, there's a number of benefits in being able to encapsulate a stand-alone C# app in 1 .cs which I've [already mentioned](https://github.com/dotnet/csharplang/issues/3502#issuecomment-632608536), notwithstanding being able to multiple stand-alone .cs programs in a folder, which currently requires a different folder + .csproj per App.

> Just because people don't like your idea and bring up issues with it does not mean that their issues are wrong.

Well it has more likes than dislikes and the only push back is from people like yourself who seemingly have no idea (or interest) in the stated goals & benefits of what's being requested, are instead adding negative value to the discussion with false statements and equivalences. 

> TLDR: welcome to language design :)

If that's the case I'll be sure not to make the mistake of wasting my time by posting here ever again.

**Comment by YairHalberstadt** [comment]
Csharp has no idea msbuild exists. The compiler (CSC) is just a command line program that takes a list of files and a list of references and compiles them.

Msbuild is a build system that is responsible for calling CSC with the right arguments. It's constantly changing and updating.

As such I don't think it makes sense to give csharp special knowledge of msbuild.

It would however be perfectly reasonable for msbuild to take a file like you requested, strip the XML, and pass that to CSC.

As such I think this request belongs on the msbuild repo.

**Comment by mythz** [comment]
> As such I think this request belongs on the msbuild repo.

It was originally [suggested by a .NET PM](https://twitter.com/DamianEdwards/status/1263178203650076672) to open a feature request for this at dotnet/roslyn, it's since been transferred by a Roslyn developer here, but it should now be moved to `/dev/null` as I've closed the request & have no further interest in wasting any more time discussing it.

**Comment by YairHalberstadt** [comment]
Also, it would probably be trivial to write your own dotnet global tool, which takes this file, splits it in two in a temp directory, and then calls msbuild on it.

**Comment by mythz** [comment]
dotnet scripting/interactive tools already exists, if everyone needs a specialized tool to run a specialized `.cs` it defeats the purpose of having a standardized format/capability that everyone can share & run with the standard dotnet tools.

**Comment by 333fred** [comment]
> dotnet scripting/interactive tools already exists, if everyone needs a specialized tool to run a specialized .cs it defeats the purpose of having a standardized format/capability that everyone can share & run with the standard dotnet tools.

Then perhaps that makes an argument that anyone should be able to `dotnet run` a .cs file with nothing more than the standard installation, no? To be clear: this type of scenario has existing for _years_ in CSX files. We know how it works (quite well, if csx were actually supported in more places). Adding references via `#r` is easy and, importantly, consistent with other compiler `#` directives. Much of the setup that a csproj would provide is either not particularly necessary in a single-file program or would be verbose enough that most developers would want to start separating out concerns in order to make reading and reasoning about code simpler. As an example, a csproj talks about:

* The output type of the program. It's a single .cs file with top-level statements. If it's not exe, it's going to fail to compile. That can just be assumed.
* The target framework: Well, that's going to depend on what version of `dotnet` you use to run it, and would almost always be sensible to just assume `latest`. If you don't want the latest version, well at that point you don't want a straightforward simple .cs file.
* Dependency expression: We have an existing experience here around `#r`. We have existing tooling that works quite well, the C# parser requires no changes in order to understand it, and much of full support in .cs files would kinda just fall out of existing tooling.
* Assembly naming and versioning: These things just generate an `AssemblyVersion.cs` file with the appropriate global attributes filed in. You could just put those attributes in your file like you can today, no issues.

Anything beyond those simple steps, and now you don't really have a simple program anymore. Seems like a perfectly reasonable cliff to fall off to a full-blown project to me, all using things that C# developers are familiar with today.

> CyrusNajmabadi You're being especially obtuse and seemingly have no clue of what's being requested.

Cyrus has every idea of what is being requested. He has just expressed legitimate concerns with it. I agree with his concerns. Different people are allowed to have different opinions on the merits of an idea.

> Well it has more likes than dislikes and the only push back is from people like yourself who seemingly have no idea (or interest) in the stated goals & benefits of what's being requested, are instead adding negative value to the discussion with false statements and equivalences.

Both Cyrus and myself have every interest in making single-file C# scripts easier to run. We just disagree with you on the best method by which to accomplish this. These are not false equivalences, they are legitimate points. I share Cyrus's opinion that adding an entirely separate language inside a C# file is a massive cognitive load that is both confusing and hard to understand. Also, as has been pointed out, embedding C# inside MSBuild is possible today. Is it more verbose? Yes, it is, in that raw example. But it would be extremely possible to make an MSBuild SDK that could shorten the example to something like this:

```xml
<Project Sdk="Microsoft.NET.EmbededCSharp.SDK">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <InlineCode>

		System.Console.WriteLine("Hello World!");

    </InlineCode>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="AwesomeSauce" Version="1.*" />
  </ItemGroup>
</Project>
```

This is practically the same number of lines as your original example, and would not be particularly difficult of an SDK to make (and would certainly be significantly less work than trying to do it the other way around).

Finally, I'll ask you to please read the .NET Foundation Code of Conduct, which this repo abides by: https://dotnetfoundation.org/about/code-of-conduct. Multiple people (including now 2 language designers) have given you feedback on your proposal. Our feedback has been good faith, legitimate criticisms. I would ask that you accept them in the same fashion.

**Comment by mythz** [comment]
I've already closed this feature request, what else do I have to do? Seriously ignore that I've ever posted on here & open up your own feature request with your own stated goals. I've already explained how [C# in MSBuild](https://github.com/dotnet/csharplang/issues/3502#issuecomment-635053660) is measurably worse in every desired goal & scenario my feature request was meant to enable. 

Whilst this proposal may make it easier to embed C# in MSBuild than what already exists, this is in't better or more appealing for new comers or being able to share encapsulated `.cs` top-level programs. Examples of uses cases I would've liked to use stand-alone `.cs` for:

 - maintain multiple unix bash scripts in a directory as I do now with `.sh` and (#Script .ss,.sc,.l ext)
 - share stand-alone runnable `.cs` examples (either for showing how to use libs or create repros)

I've no desire to use "C# in MSBuild" to do any of these things, the only time I'd ever consider embedding C# in MSBuild XML is to enable functionality that doesn't exist in MSBuild, last time I needed to do that was to workaround MSBuild not having a non-blocking Exec task.

I find it hard to believe the same language designers that brought C# 9's UX-friendly top-level programs somehow finds the more verbose & ugly & heavily indented lets "embed C# in MSBuild XML" a more suitable proposal. If you want to enable this, open up a feature request and optimize it for whatever state use-cases that it's supposed to be more desirable for, as it's not useful for anything I was hoping to use it for.

**Comment by orthoxerox** [comment]
@333fred 

The only thing a single file CSX can't do is indicate it wants the Web SDK instead of the regular one. `#sdk`?

**Comment by YairHalberstadt** [comment]
@333fred 

The biggest limitation of CSX is the inability to reference nuget packages.

**Comment by HaloFour** [comment]
@YairHalberstadt 

`dotnet script` supports Nuget references in `#r` directives:

https://galdin.dev/blog/csharp-scripts-using-dotnet-script/

**Comment by svick** [comment]
@orthoxerox Are single-file websites a common requirement? Even `dotnet new web` creates two separate .cs files.

@YairHalberstadt That depends on the script runner. For example, [`dotnet script` supports it](https://github.com/filipw/dotnet-script#nuget-packages-1).

**Comment by orthoxerox** [comment]
@svick they demoed it at Build

**Comment by jnm2** [comment]
I think the talk in question is https://channel9.msdn.com/Events/Build/2020/BOD106 at 52m45s.

**Comment by CyrusNajmabadi** [comment]
> I find it hard to believe the same language designers that brought C# 9's UX-friendly top-level programs somehow finds the more verbose & ugly & heavily indented lets "embed C# in MSBuild XML" a more suitable proposal.

I do not know what to tell you.  We're the language designers I've been working in this space before 2000, and i started right after C# 1 and have helped design and ship and implement almost every version after that.  

Our feedback was made in good faith with a ton of experience and knowledge about these parts of the system.  Perhaps consider then that we share a different set of values and beliefs than you *and that's ok*.  C# and .net is not a homogeneous mass of people all operating in lockstep on what they think is right.  It's the culmination of literally billions of man hours over decades.  Proposals are ultimate going to be subjected to the scrutiny and complex value systems that have come out of that.  

>  I've already explained how C# in MSBuild is measurably worse in every desired goal & scenario my feature request was meant to enable.

Here's the thing.  That may be *your* assessment.  It may not be a belief shared by others.  *Furthermore*, even if your assessment is *correct* and believed in by others, that doesn't mean your *solution* is felt to be the most appropriate one.  It may be the one you *want* teh most, but that doesn't automatically make it the best, or the one *others* believe would be best to make this space better. 

Language design is about rolling around many options and making many tradeoffs to come up with solutions that we think can deliver solutions we think fit a ton of difficult constraints.  So while you like your solution based on the things you care (and don't care) about, we have to look broadly at tons of other criteria that we know are still important to the ecosystem as a whole.

--

Finally, let me say this.  We've spent the time to read and understand and engage with you civilly.  I believe i understand your position and can commiserate with it.  The issues you want to solve are certainly there.  I just disagree on your solution being the right path forward.  Please do us the courtesy of attempting to understand and see our perspective on things.  It will be very helpful, and will likely allow any further proposals of yours in the future to go a lot further.  Thanks!

**Comment by TAGC** [comment]
Aight how do people feel about this:

You know how in NuGet you can specify different NuGet.Config files to use when pulling packages (etc.), and if you don't specify anything it defaults to one at a well-known location on the file system?

What if you had a default location on the OS (e.g. `%AppData%\dotnet`) that dotnet tries to locate the MSBuild project information at if you pass a special flag to `dotnet run`. Like @333fred says, for most cases of single-file C# scripts I'm going to be using the same settings: compiling to an executable, targeting the latest framework, etc.

As for project dependencies, I don't care if this "default csproj" (so to speak) ends up bloated with lots of package references since it's the same deal with Python: unless you use virtual environments, you tend to rack up dozens of packages on your system as you write different scripts and it doesn't cause much issue to just leave them there unless you specifically need to work with different versions of the same dependency in different scripts, which in practice hasn't happened for me all that often.

**Comment by olfek** [comment]
Be practical everyone ...

We want:
```
dotnet run main.cs
```

**Without** needing to specify any package references.

 1. dotnet in a script mode parses the `using` statements and looks for relevant packages in the global `.nuget` packages folder.
 1. dotnet passes `main.cs` to the Roslyn compiler along with references to the relevant packages found in the global `.nuget` packages folder.
 1. `dotnet` program runs executable emitted by Roslyn compiler.
 1. Profit

**Comment by olfek** [comment]
Can you reopen this please @mythz 

We really need this for C#

I've been using Dart for scripts. `dart run <filename>`

**Comment by mythz** [comment]
@olfek No point it's clear the C# team have no interest in supporting this use case. 

I've switched to `node script.js`, modern JavaScript is quite a pleasant ergonomic wrist-friendly language with fast startup times that's IMO better suited for creating & running scripts.

**Comment by olfek** [comment]
@mythz I created this ➡️ https://github.com/dotnet/csharplang/discussions/7552
